test: Notifications Inbox Screen Unit Tests Implementation

shafqat-muneer opened the pull request:
[LEARNER-10560](https://2u-internal.atlassian.net/browse/LEARNER-10560): Notifications Module Unit Tests: Inbox Screen

Adds test cases to improve coverage and ensure stability for the Notification Inbox Screen. These tests validate expected behaviors and edge cases across different scenarios.

Commit:
test: Notifications Inbox Screen Unit Tests Implementation

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+                lastRead: nil,
+                lastSeen: Date(iso8601: "2025-01-06T01:20:58.919612Z"),
+                created: Date(iso8601: "2025-01-06T01:20:58.919612Z")
+            ), SingleNotification(
```
better to start new Notification with new line

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+            notificationsInteractor: mockInteractor,
+            analytics: NotificationsAnalyticsMock(),
+            router: NotificationsRouterMock(),
+            connectivity: Connectivity(),
```
shouldn't we use here `mockConnectivity`?

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+        super.tearDown()
+    }
+    
+    private func makeSUT(
```
where do you use this function?

shafqat-muneer replied:
It was written to use it but later there were not any case. Just removed for now. Thanks for highlighting.

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+    func testGetAllNotificationsSuccess() async throws {
+        Given(mockInteractor, .getAllNotifications(page: 1, willReturn: notifications))
+        
+        viewModel = NotificationsInboxViewModel(
```
Why do we need to create viewModel once again if we did it in setUp?

shafqat-muneer replied:
Intention behind was to execute full functionality at a single test case level but you are also right. For now, I just removed re-initialization from single test case level.

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+        
+        await viewModel.markNotificationAsRead(notificationId: "1")
+        
+        Verify(mockInteractor, .markNotificationAsRead(notificationId: .any))
```
Lets add `count` parameter here (how much times func was called) - and do this for all tests?
Also if we called `markNotificationAsRead` for Notification with id="1" then it makes sense to `Verify` this value but not `.any` (same in another tests).
Also let's check we don't have error with `XCTAssertNil(viewModel.errorMessage)`

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+    }
+    
+    func testMarkNotificationAsReadUnknownError() async throws {
+        let noInternetError = AFError.sessionInvalidated(error: NSError(domain: "error", code: -1, userInfo: nil))
```
bad let name for `unknown` error

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+        
+        Verify(mockInteractor, .markNotificationAsRead(notificationId: .any))
+        
+        XCTAssertNotNil(viewModel.errorMessage)
```
if we just check `errorMessage != nil` then this test does the same as next one (testMarkNotificationAsReadUnknownError) because `errorMessage` gets value for any throwing error.
Maybe it's better to check if `errorMessage` got particular error message (`CoreLocalization.Error.slowOrNoInternetConnection` in this test)?
WDYT?

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+        
+        Verify(mockInteractor, .markAllNotificationsAsRead())
+        
+        XCTAssertNotNil(viewModel.flatNotifications[0].lastRead)
```
Maybe we need to check all notifications, not only one?

rnr replied:
Also will be good to check if errorMessage is `nil`

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+        
+        Verify(mockInteractor, .markAllNotificationsAsRead())
+        
+        XCTAssertNotNil(viewModel.errorMessage)
```
same as for `testMarkNotificationAsReadNoInternetError`

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+        
+        await viewModel.markNotificationsAsSeen()
+        
+        Verify(mockInteractor, .markNotificationsAsSeen())
```
As I see errors are not catching for `markNotificationsAsSeen`. So what we are testing in this and next test?

shafqat-muneer replied:
I think, it's a valid and meaningful test. We’re not testing the result or state, but we are verifying system behavior that the interactor is called even under error conditions — and the system remains stable.

rnr replied:
ok, but why we need to do it twice (for `no internet` and `unknown` errors) if it doesn't depend on error type?

shafqat-muneer replied:
Makes sense — since the behavior is the same for both errors, I’ve remove one of the test cases to avoid redundancy.

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+    }
+    
+    func testGetAllNotificationsSuccess() async throws {
+        Given(mockInteractor, .getAllNotifications(page: 1, willReturn: notifications))
```
We did this in setUp already, right?

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+        Given(mockInteractor, .getAllNotifications(page: 1, willThrow: URLError(.notConnectedToInternet)))
+        Given(mockConnectivity, .isInternetAvaliable(getter: false))
+        
+        viewModel = NotificationsInboxViewModel(
```
same here

shafqat-muneer replied:
https://github.com/edx/edx-mobile-marketplace-ios/pull/148#discussion_r2140776025

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+        Given(mockInteractor, .getAllNotifications(page: 1, willThrow: NSError(domain: "error", code: -1, userInfo: nil)))
+        Given(mockConnectivity, .isInternetAvaliable(getter: true))
+        
+        viewModel = NotificationsInboxViewModel(
```
same here

shafqat-muneer replied:
https://github.com/edx/edx-mobile-marketplace-ios/pull/148#discussion_r2140776025

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,309 @@
+        XCTAssertEqual(viewModel.flatNotifications[0].appName, "discussion")
+    }
+    
+    func testNotificationTypeAsCommentOnFollowedPost() async throws {
+        await viewModel.loadNotifications()
+        
+        XCTAssertEqual(viewModel.flatNotifications[0].notificationType, "comment_on_followed_post")
+    }
+    
+    func testNotificationTypeAsResponseOnFollowedPost() async throws {
+        await viewModel.loadNotifications()
+        
+        XCTAssertEqual(viewModel.flatNotifications[1].notificationType, "response_on_followed_post")
+    }
+    
+    func testNotificationTypeAsNewResponse() async throws {
+        await viewModel.loadNotifications()
+        
+        XCTAssertEqual(viewModel.flatNotifications[2].notificationType, "new_response")
+    }
+    
+    func testNotificationTypeAsNewComment() async throws {
+        await viewModel.loadNotifications()
+        
+        XCTAssertEqual(viewModel.flatNotifications[3].notificationType, "new_comment")
+    }
```
Do you think we need all these tests? Which NotificationsInboxViewModel's function we test here? 
By fact we just check if string from mock let `notifications` is equal to another string.

rnr commented on Notifications/Notifications/Presentation/NotificationsInbox/NotificationsInboxViewModel.swift:
```diff
@@ -31,7 +31,7 @@ public class NotificationsInboxViewModel: ObservableObject {
     private var cancellables = Set<AnyCancellable>()
     private(set) var lazyVStackUniqueID: String = UUID().uuidString
-    private var flatNotifications: [SingleNotification] = [] {
+    var flatNotifications: [SingleNotification] = [] {
```
`private(set)` is better then

mta452 reviewed (changes requested):
Here’s my general feedback on this:
* It would be better to use JSON string directly and decode it into respective data types if possible.
* It would be better to use `makeSUT(...)` for preparing the relevant variables in each test case instead of relying on general `setUp()` method. Also, renaming `viewModel` to either `sut` or `subject` would clearly indicate the core data type being tested.
* Assertions for supporting side effects should also be added in each test case.
* Test cases for menu selection can be added to verify proper api calls, navigation, analytics etc.
* Test cases for load/refresh/fetchMore notifications should be added to verify the proper screen states as well as the notifications data.
* A test case to check the proper grouping of notifications would be really helpful to catch unexpected future bugs around this.
* I remember we struggled with an issue where scrolling to the bottom of the list wasn’t triggering the load more operation. A test case would be very helpful to ensure that the infinite scrolling behavior remains intact.

Commit:
fix: address review feedback

shafqat-muneer commented:
> * It would be better to use JSON string directly and decode it into respective data types if possible.

Right now, this approach is being used throughout the code, and we're focusing on testing the `NotificationsInboxViewModel` functionality. We can write test cases for the decoding part later, as part of a separate ticket.

> * It would be better to use `makeSUT(...)` for preparing the relevant variables in each test case instead of relying on general `setUp()` method. Also, renaming `viewModel` to either `sut` or `subject` would clearly indicate the core data type being tested.

- Renamed viewModel to sut.
- Using `makeSUT()` instead of `setUp()` is a great suggestion. However, in our current tests, we’re using the same mocks and setup across all cases, and we're not modifying the `viewModel` dependencies per test. So, at this stage, using `setUp()` seems appropriate. We can revisit this and switch to `makeSUT()` later if we need to override behavior in specific test cases.

> * Assertions for supporting side effects should also be added in each test case.

If you're referring to `XCTAssert` comparisons, I've added a few more of them.

> * Test cases for menu selection can be added to verify proper api calls, navigation, analytics etc.
> * Test cases for load/refresh/fetchMore notifications should be added to verify the proper screen states as well as the notifications data.
> * A test case to check the proper grouping of notifications would be really helpful to catch unexpected future bugs around this.
> * I remember we struggled with an issue where scrolling to the bottom of the list wasn’t triggering the load more operation. A test case would be very helpful to ensure that the infinite scrolling behavior remains intact.

These would be a nice addition to the test cases. Our intention was to get started with some basic tests and gradually build on them. We can enhance and add more test cases as part of a separate ticket.

shafqat-muneer commented:
@rnr @mta452 I've addressed and commented on the review feedback. The PR is ready for another round of review. Thanks!

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,259 @@
+        
+        Verify(mockInteractor, 1, .markNotificationAsRead(notificationId: .value("1")))
+        
+        XCTAssertNotNil(sut.errorMessage)
```
next line will fail if errorMessage is `nil`. Do we need both?

shafqat-muneer replied:
I’ve added the first assertion to explicitly confirm that errorMessage is set, and the second to verify it matches `.slowOrNoInternetConnection`. This helps isolate whether the failure is due to a missing error or an unexpected value, making the test output clearer. Let me know if you’d prefer simplifying it.

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,259 @@
+        
+        Verify(mockInteractor, 1, .markAllNotificationsAsRead())
+        
+        XCTAssertNotNil(sut.errorMessage)
```
the same: next line will fail if errorMessage is `nil`

shafqat-muneer replied:
https://github.com/edx/edx-mobile-marketplace-ios/pull/148#discussion_r2149787911

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,259 @@
+        await sut.loadNotifications()
+        await sut.markAllNotificationsAsRead()
+        
+        Verify(mockInteractor, 1, .markAllNotificationsAsRead())
```
maybe we need to check `lastRead` is still `nil`? like `XCTAssertNil(sut.flatNotifications[0].lastRead)`

shafqat-muneer replied:
Added against mock elements.

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,259 @@
+        await sut.loadNotifications()
+        await sut.markAllNotificationsAsRead()
+        
+        Verify(mockInteractor, 1, .markAllNotificationsAsRead())
```
same `XCTAssertNil(sut.flatNotifications[0].lastRead)` here?

shafqat-muneer replied:
Added against mock elements.

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,259 @@
+    func testGetAllNotificationsSuccess() async throws {
+        await sut.loadNotifications()
+        
+        XCTAssertEqual(notifications.results?.count, sut.flatNotifications.count)
```
maybe add verification of calls count for `.getAllNotifications`?

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,259 @@
+        
+        await sut.loadNotifications()
+        
+        XCTAssertEqual(sut.screenState, .noInternet)
```
check `sut.flatNotifications.count` is equal `zero` here?
and verification of calls count for .getAllNotifications?

rnr commented on Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:
```diff
@@ -0,0 +1,259 @@
+        
+        await sut.loadNotifications()
+        
+        XCTAssertEqual(sut.screenState, .serverError)
```
same here

rnr reviewed (changes requested):
Thanks for the good work. Just a few more questions )

Commit:
fix: address review feedback

shafqat-muneer commented:
@rnr I've addressed and commented on the review feedback. Kindly have another round of review. Thanks!

rnr reviewed (approved):
LGTM. Thank you

mta452 reviewed (approved):
The implemented test cases look good, would be best if suggested test cases are also added.

shafqat-muneer commented:
> The implemented test cases look good, would be best if suggested test cases are also added.

These would be a nice addition to the test cases. Our intention was to get started with some basic tests and gradually build on them. We can enhance and add more test cases as part of a separate ticket.
